core/src/main/java/org/apache/iceberg/MergingSnapshotUpdate.java

```diff
@@ -355,80 +351,110 @@ private ManifestFile filterManifest(Expression deleteExpression,
-      try {
-        for (ManifestEntry entry : reader.entries()) {
-          DataFile file = entry.file();
-          boolean fileDelete = (deletePaths.contains(pathWrapper.set(file.path())) ||
-              dropPartitions.contains(partitionWrapper.set(file.partition())));
-          if (entry.status() != Status.DELETED) {
-            if (fileDelete || inclusive.eval(file.partition())) {
-              ValidationException.check(
-                  fileDelete || strict.eval(file.partition()) || metricsEvaluator.eval(file),
-                  "Cannot delete file where some, but not all, rows match filter %s: %s",
-                  this.deleteExpression, file.path());
-
-              writer.delete(entry);
-
-              CharSequenceWrapper wrapper = CharSequenceWrapper.wrap(entry.file().path());
-              if (deletedPaths.contains(wrapper)) {
-                LOG.warn("Deleting a duplicate path from manifest {}: {}",
-                    manifest.path(), wrapper.get());
-                summaryBuilder.incrementDuplicateDeletes();
-              } else {
-                // only add the file to deletes if it is a new delete
-                // this keeps the snapshot summary accurate for non-duplicate data
-                deletedFiles.add(entry.file().copy());
-              }
-              deletedPaths.add(wrapper);
+      return filterManifestsWithDeletedFiles(
+          metricsEvaluator,
+          manifest,
+          reader,
+          inclusive,
+          strict,
+          pathWrapper,
+          partitionWrapper);
+    }
+  }
 
+  private boolean manifestHasDeletedFiles(
+      StrictMetricsEvaluator metricsEvaluator,
+      ManifestReader reader,
+      Evaluator inclusive,
+      Evaluator strict,
+      CharSequenceWrapper pathWrapper,
+      StructLikeWrapper partitionWrapper) {
+    boolean hasDeletedFiles = false;
+    for (ManifestEntry entry : reader.entries()) {
+      DataFile file = entry.file();
+      boolean fileDelete = deletePaths.contains(pathWrapper.set(file.path())) ||
+          dropPartitions.contains(partitionWrapper.set(file.partition()));
+      if (fileDelete || inclusive.eval(file.partition())) {
+        ValidationException.check(
+            fileDelete || strict.eval(file.partition()) || metricsEvaluator.eval(file),
+            "Cannot delete file where some, but not all, rows match filter %s: %s",
+            this.deleteExpression, file.path());
+
+        hasDeletedFiles = true;
+        if (failAnyDelete) {
+          throw new DeleteException(writeSpec().partitionToPath(file.partition()));
+        }
+        break; // as soon as a deleted file is detected, stop scanning
+      }
+    }
+    return hasDeletedFiles;
+  }
+
+  private ManifestFile filterManifestsWithDeletedFiles(
+      StrictMetricsEvaluator metricsEvaluator,
+      ManifestFile manifest,
+      ManifestReader reader,
+      Evaluator inclusive,
```

Comment:
If these evaluators are cached, can this method retrieve them instead of passing them in?

Reply:
Not quite sure I follow. The evaluators are also passed into `manifestHasDeletedFiles`. How do I retrieve them from the cache?

Reply:
I just would like to avoid such a huge argument list. If these evaluators are fetched from a cache based on the reader's partition spec, then those fetches could happen here instead to avoid so many arguments.

Reply:
These evaluators are used both here and in the other helper method, `manifestHasDeletedFiles`. Hence why we evaluate them at the level above and then pass through method args.